File: day04/main.cpp
#include <iostream>
#include <fstream>
#include <vector>
#include <algorithm>
#include <numeric>
#include <sstream>
#include  <regex>
#include <optional>
#include <functional>
#include <set>
#include <unordered_map>
#include <map>

std::set<int64_t> find_integers(const std::string& str) {
    static std::regex base_regex("(\\d+)");

    std::set<int64_t> result;
    for (auto it = std::sregex_iterator(str.begin(), str.end(), base_regex);
         it != std::sregex_iterator(); ++it) {

        auto value = std::stoi(it->operator[](1).str());
        result.insert(value);
    }
    return result;
}

void part1() {

    std::fstream ifs(std::string(SRC_DIR)+"/input.txt");
    std::string line;

    int64_t total = 0;
    while(std::getline(ifs, line)) {

        auto colon = line.find(":");
        auto sep = line.find("|");

        auto winning_numbers = find_integers(line.substr(colon+1, sep-colon));
        auto card_numbers = find_integers(line.substr(sep+1));

        std::vector<int64_t> common;
        std::set_intersection(winning_numbers.begin(), winning_numbers.end(), card_numbers.begin(), card_numbers.end(), std::back_inserter(common));
        if(!common.empty()) {
            total += (1 << (common.size() - 1));
        }
    }

    std::cout << "part1: " << total << std::endl;
}

void part2() {

    std::fstream ifs(std::string(SRC_DIR)+"/input.txt");
    std::string line;

    std::vector<int64_t> matching_numbers;

    while(std::getline(ifs, line)) {

        auto colon = line.find(':');
        auto sep = line.find('|');

        auto winning_numbers = find_integers(line.substr(colon+1, sep-colon));
        auto card_numbers = find_integers(line.substr(sep+1));

        std::vector<int64_t> common;
        std::set_intersection(winning_numbers.begin(), winning_numbers.end(), card_numbers.begin(), card_numbers.end(), std::back_inserter(common));
        matching_numbers.push_back(common.size());
    }



    std::vector<int64_t> total_cards(matching_numbers.size(), 1); // original card

    for(int i=0; i < matching_numbers.size(); ++i) {
        auto count = total_cards[i];
        auto matching = matching_numbers[i];
        for(int j=i+1; j < i+1+matching; ++j) {
            total_cards[j]+=count;
        }
    }

    int64_t total = std::accumulate(total_cards.begin(), total_cards.end(), 0ll);

    std::cout << "part2: " << total << std::endl;
}

int main() {
    part1();
    part2();
    return 0;
}

File: day04/src/main.rs
use std::collections::HashSet;
use lazy_static::lazy_static;
use regex::{Match, Regex};
use std::fs::File;
use std::io::{self, BufRead};
use std::iter::Map;
use std::path::Path;
use std::slice::Iter;

fn read_lines<P>(filename: P) -> io::Result<io::Lines<io::BufReader<File>>>
where
    P: AsRef<Path>,
{
    let file = File::open(filename)?;
    Ok(io::BufReader::new(file).lines())
}

fn read() -> Vec<String> {
    let file_path = "input.txt";
    let Ok(lines) = read_lines(file_path) else {
        return vec!();
    };

    return lines.into_iter().map(|x| x.unwrap()).collect();
}

fn find_integers(input: &str) -> HashSet<i64> {

    lazy_static! {
        static ref NUMBER_RE: Regex = Regex::new(r#"(?P<number>[0-9]+)"#).unwrap();
    }

    let mut result: HashSet<i64> = HashSet::new();
    for capture in NUMBER_RE.captures_iter(input) {
        let number = capture.name("number").unwrap().as_str().parse::<i64>().unwrap();
        result.insert(number);
    }
    result
}

fn part1() {

    let mut total = 0i64;
    for line in read() {

        let start = line.find(':').unwrap();
        let delimiter = line.find('|').unwrap();

        let winning_numbers = find_integers(&line[start+1..delimiter]);
        let card_numbers = find_integers(&line[delimiter+1..]);

        let matched = winning_numbers.intersection(&card_numbers).count();
        if matched > 0 {
            total += 1 << (matched - 1);
        }
    }

    println!("part1 {}", total);
}

fn part2() {

    let mut matching_numbers: Vec<i64> = vec![];
    for line in read() {

        let start = line.find(':').unwrap();
        let delimiter = line.find('|').unwrap();

        let winning_numbers = find_integers(&line[start+1..delimiter]);
        let card_numbers = find_integers(&line[delimiter+1..]);

        let matched = winning_numbers.intersection(&card_numbers).count();
        matching_numbers.push(matched as i64);
    }

    let mut total_cards: Vec<i64> = vec![1; matching_numbers.len()];

    for idx in 0..total_cards.len() {
        let count = total_cards[idx];
        let matched = matching_numbers[idx] as usize;
        for j in idx+1..idx+1+matched {
            total_cards[j]+=count;
        }
    }

    let total: i64 = total_cards.iter().sum();
    println!("part2 {}", total);
}

fn main() {
    part1();
    part2();
}
